File: biglist_filter_kentekens.pl
#!/usr/bin/env perl

### Test program

use Text::CSV_XS;
use IO::File;

$input_file = "./Open_Data_RDW__Gekentekende_voertuigen.csv";
$input_size =  -s $input_file;

@output_files = ();
@paths = ();

if (! open ENTREE, "<", $input_file) {
	print "Could not open handle ENTREE for $input_file. $!\n";}


$start = time();


# my $csv_reader = Text::CSV_XS->new();
# while (my $row = $csv_reader->getline (<ENTREE>)) {
# 	print "$row->[2]\t$row->[0]\n";
# }
# $csv->eof or $csv->error_diag ();

sub openOutputFile {
	my $index = shift;
	my $file_name = "./Biglist_Outputs/$index.csv";
	my $fh = new IO::File "> $file_name" or die "Cannot open $file_name: $!";
	push @output_files, $fh;
}

sub writeOutputLine {
	my $index = shift;
	my $line = shift;
	my $outputFile = $output_files[$index];
	print {$outputFile} $line;
}

sub closeOutputFiles {
	foreach(@output_files) {
		close $_
	}
}

sub elapsed {
	return time() - $start;
}

sub percentage {
	return 100 * (tell ENTREE) / $input_size;
}

sub remaining {
	my $percentage = percentage();
	return (100 - $percentage) * elapsed() / $percentage;
}

sub totalTime {
	return elapsed() + remaining();
}

sub ncw_time_formatter
{
    my $seconds = shift;
    my $string = join ":", map { sprintf "%02d", $_} (gmtime($seconds))[7,2,1,0]; 
    $string=~s/\G00://g;
    return $string;
}

sub update_status_output
{
	$elapsed = ncw_time_formatter(elapsed());
	$percentage = sprintf('%.2f',percentage());
	$remaining = ncw_time_formatter(remaining());
	$total = ncw_time_formatter(totalTime());
	print "\e[K\r$percentage%\tElapsed:$elapsed\tRemaining:$remaining\tTotal Expected Time: $total";
}


my $csv_reader = Text::CSV_XS->new();
my @columns;
my $row = 0;
my $col = 0;

while (<ENTREE>) {
	$csv_reader->parse($_);
	@columns = $csv_reader->fields();
	$col = -1;
	if( $row == 0 ) {
		foreach(@columns) {
			$col++;
			openOutputFile($col);
			writeOutputLine($col, "$columns[0],$_\n");
		}
	} else {
		foreach(@columns) {
			$col++;
			if($_ eq "") {
				next;
			}
			writeOutputLine($col, "$columns[0],$columns[$col]\n");
		}
	}
	if( $row % 3) {
		update_status_output();
	}
	if( $row > 100000 ) {
		last;
	}
	$row++;
}

$duration = elapsed();

print "\nReading took $duration seconds.\n";
closeOutputFiles;
close ENTREE;

File: biglist_filter_kentekens.php
#!/usr/bin/env php
<?php
const BASE_DIR = __DIR__ . '/';
const INPUT_FILE = BASE_DIR . 'Open_Data_RDW__Gekentekende_voertuigen.csv';
const OUTPUT_DIR = BASE_DIR . 'Biglist_Outputs/';
const INDEX_FILE = OUTPUT_DIR . 'index.csv';


const DEFAULT_BLOCK_SIZE = 8192;
const BLOCK_SIZE = 16 * DEFAULT_BLOCK_SIZE;

const CLN = "\033[2K\r";

class IO_FILE {
	private $fp;
	private $filename;

	public function getFilename() { return $this->filename; }
	public function getFileHandle() { return $this->fp; }

	protected function _open($filename, $mode) {
		$this->filename = $filename;
		$this->fp = fopen($filename, $mode);

		$this->_confirm_open();
	}

	protected function _filesize() {
		return filesize($this->filename);
	}

	protected function _tell() {
		return ftell($this->fp);
	}

	protected function _confirm_open() {
		if(!$this->getFileHandle()) throw new Exception('File not open... '.$this->getFilename());
	}

	protected function _write($data) {
		return fwrite($this->fp, $data);
	}

	protected function _gets() {
		return fgets($this->fp);
	}

	protected function _eof() {
		return feof($this->fp);
	}

	protected function _read($length) {
		return fread($this->fp, $length);
	}

	protected function _close() {
		fclose($this->fp);
	}

	public function __destruct() {
		$this->_close();
	}
}

class OutputFile extends IO_FILE{
	private $buffer = '';

	public function __construct($filename) {
		$this->_open($filename, 'w');
	}

	private function flush() {
		$result = $this->_write($this->buffer);
		if($result === false) throw new Exception('File Write Error:'.$this->getFilename());
		$this->buffer = substr($this->buffer, $result);
	}

	protected function writeLine($data) {
		if(strlen($this->buffer) > BLOCK_SIZE) $this->flush();
		$this->buffer .= $data.PHP_EOL;
	}

	public function __destruct() {
		while(strlen($this->buffer) > 0) {
			$this->flush();
		}
		parent::__destruct();
	}
}

class indexFile extends OutputFile {
	private $counter = 0;

	public function add($header) {
		$this->writeLine($this->counter.':'.$header);
		$this->counter ++;
	}

}
class columnFile extends OutputFile {
	public function addRow($key, $value) {
		$this->writeLine($key.','.$value);
	}
}

class InputFile extends IO_FILE {
	private $headers = [];

	public function __construct($filename) {
		$this->_open($filename, 'r');
	}

	public function getHeaders() {
		if($this->headers !== []) return $this->headers;

		$headerLine = $this->_gets();

		$this->headers = str_getcsv($headerLine);
		return $this->headers;
	}

	public function fractionalPosition() {
		return $this->_tell() / $this->_filesize();
	}

	public function __invoke($block_size = BLOCK_SIZE) {
		$this->getHeaders();

		$left='';
		while (!$this->_eof()) {// read the file

			$temp = $this->_read($block_size);
			$fgetslines = explode(PHP_EOL, $temp);
			$fgetslines[0] = $left.$fgetslines[0];

			if(!$this->_eof()) $left = array_pop($fgetslines);

			foreach($fgetslines as $k => $line){
				yield $line;
			}
		}
	}
}

class TimeKeeper {
	private $start;
	private $lastProjection = 0;

	public function __construct() {
		$this->start = microtime(true);
	}

	public function elapsed() {
		return microtime(true) - $this->start;
	}

	public function projection($percentage) {
		return $this->lastProjection = $this->elapsed() / $percentage * (100 - $percentage);
	}

	public function totalTime() {
		return $this->lastProjection + $this->elapsed();
	}
}

class ProgressDisplay extends TimeKeeper{
	private $percentage = 0.0001;

	public function __invoke($percentage) {
		if($percentage > 0)
			$this->percentage = $percentage;
	}

	public function timeElapsed() {
		return $this->secondsToTime($this->elapsed());
	}

	public function timeRemaining() {
		return $this->secondsToTime($this->projection($this->percentage));
	}

	public function timeTotal() {
		return $this->secondsToTime($this->totalTime());
	}

	public function outputStatusLine($percentage = 0) {
		$this($percentage);
		echo CLN
		, 'Processing... ('
		, $this->decimalPercentage($percentage)
		, '%) Elapsed:'
		, $this->timeElapsed()
		, ' Remaining:'
		, $this->timeRemaining()
		, ' Total:'
		, $this->timeTotal();
	}

	private function secondsToTime($s) {
		$h = floor($s / 3600);
		$s -= $h * 3600;
		$m = floor($s / 60);
		$s -= $m * 60;
		return $h.':'.sprintf('%02d', $m).':'.sprintf('%02d', $s);
	}

	private function decimalPercentage($p) {
		return sprintf('%.02f', $p);
	}
}

class ColumnFiler {
	private $files = [];
	private $outputDir;

	private $numColumns = 0;

	public function __construct($outputDir = OUTPUT_DIR) {
		if(!file_exists($outputDir)) throw new Exception('Output Directory Does Not Exist:'.$outputDir);
		$this->outputDir = $outputDir;
	}

	public function registerOutputFiles($headers) {
		$index = new indexFile($this->outputDir.'index.csv');

		foreach($headers as $n => $type) {
			$index->add($type);
			$this->files[$n] = new columnFile($this->outputDir.$n.'.csv');
			$this->files[$n]->addRow('Kenteken', $type); // FIXME: Un-Hardcode This!!!
		}

		$this->numColumns = $n;
	}

	public function process($fields) {
		foreach($fields as $index => $field) {
			if($field === '') continue;
			// Re-escape quotes and commas
			if(strpos($field, ',') > -1) {
				$field = str_replace('"', '\\"', $field);
				$field = '"'.$field.'"';
			}
			$this->files[$index]->addRow($fields[0], $field);
		}
	}
}

class Orchestrator {
	public function __construct() {
		$progressDisplay = new ProgressDisplay();
		$columnFiler = new ColumnFiler(OUTPUT_DIR);
		$in = new InputFile(INPUT_FILE);

		$columnFiler->registerOutputFiles($in->getHeaders());

		foreach($in() as $n => $line) {	
			$columnFiler->process(str_getcsv($line));
			if($n % 1024 === 0) $progressDisplay->outputStatusLine($in->fractionalPosition() * 100);
			if($n > 100000) {
				$progressDisplay->outputStatusLine($in->fractionalPosition() * 100);
				echo "\n";
				exit();
			}
		}

		echo "Finished Processing File. Outputs Saved. \n";
	}
}


new Orchestrator();















